### PR TITLE
Add json output similiar to what zMerge outputs

### PR DESCRIPTION
**Pull request by monyarm**
- refactor shell scripts used for testing
- disable unnecessary console logs left over from previous debugging
- fix shell expansion bug with test scripts
- Add json output similiar to what zMerge outputs

### CODE REVIEW

File: MutagenMerger.Lib/DI/CopyRecordProcessor.cs [hunk 1]
```diff
@@ -61,11 +61,17 @@ public CopyRecordProcessor(ICopyOverride<TMod, TModGetter>[] copyOverrides)
...
-        mergeState.Mapping.Add(rec.Record.FormKey, duplicated.FormKey);
+        if (rec.Record.FormKey.ID != duplicated.FormKey.ID)
+        {
+            Console.WriteLine("          Renumbering Record [" + rec.Record.FormKey.ModKey.Name + "] " +
+                              rec.Record.FormKey.IDString() + " to [" + mergeState.OutgoingMod.ModKey.Name + "] " +
+                              duplicated.FormKey.IDString());
+            mergeState.Mapping.Add(rec.Record.FormKey, duplicated.FormKey);
+        }
+        else {
+            Console.WriteLine("          Copying Record [" + rec.Record.FormKey.ModKey.Name + "] " +
```
Comment (Noggog) [inline MutagenMerger.Lib/DI/CopyRecordProcessor.cs:72]: Not a blocking change, but might consider the new string interpolation style
```
Console.WriteLine($"          Copying Record [{rec.Record.FormKey.ModKey.Name}] {rec.Record.FormKey.IDString()}");
```

File: MutagenMerger.Lib/DI/Merger.cs [hunk 2]
```diff
@@ -90,8 +92,76 @@ public sealed class Merger<TModGetter, TMod, TMajorRecord, TMajorRecordGetter> :
             HandleAssets(state);
 
             HandleScripts(state);
+
+            MergeJson(state);
         }
-        
+
+        private void MergeJson(MergeState<TMod, TModGetter> state)
```
Comment (Noggog) [inline MutagenMerger.Lib/DI/Merger.cs:99]: Two thoughts:

# Dependency Injection - Separate Class
Might be good to move this to a dedicated class, say `MergeJsonExporter`?  Since we're using dependency injection systems, you can then just list that `MergeJsonExporter` as another dependency (alongside `gameSpecs` and `copyRecordProcessor`), and then use it within `Merger`, keeping the bulk of the logic outside this file.  This helps prep for the other thought:

# Unit Tests
Do we want to write unit tests for this logic?  If these files need to be accurate/exact in order for other tools to rely on them, we might want to get some tests going for this logic at some point.   Helps to confirm we're writing these files accurately, and also helps outline the expectations.  "This file needs to have this"  "this file needs to have that", etc.

If the first thought is done, then we should have a `MergeJsonExporter` class that is just in charge of doing this exporting.  So a unit test can just make that class, and test it directly by passing in a hand crafted MergeState and checking the results, without involving the monstrous Merger superclass.

Let me know if this is something you're up to doing, and I can help give some tips on how to deal with disk IO within unit tests and all that jazz.

---------

Otherwise, looking good!  Thanks for all the hard work!

**Review by Noggog — APPROVED** [review]

File: MutagenMerger.Lib/DI/Merger.cs (hunk 2 above)
Comment (monyarm) [reply MutagenMerger.Lib/DI/Merger.cs:99]: Agreed regarding both, the json does need to follow a specific format, in order for tools like Merge Plugin Hide and Merge Mapper to recognize it. I'll see about doing this soon.

Comment (Noggog) [reply MutagenMerger.Lib/DI/Merger.cs:99]: Cool!  Here's some hopefully helpful thoughts to prep for testing, then:

# IFileSystem
If we have a `MergeJsonExporter` class, it can import `IFileSystem fileSystem` in its ctor, and use that for calls like `WriteAllToText`.   By default, it'll work exactly as before, writing to disk in normal operation, but now in unit testing, we can swap for an in-memory pretend filesystem, so none of our unit tests actually hit a disk.  Makes everything much simpler

# Autofixture
This is a nice little feature that allows unit tests to get easy randomized data, as well as basic dependency injection behaviors.   For example, a unit test might be written like so:
```cs
[Fact]
public void MyTest()
{
    var mockFs = new MockFilesystem();
    var exporter = new MergeJsonExporter(mockFs);
    
    exporter.SomeCall(...);
    
    CheckSomeFileExistsAsExpected(mockFs);
}
```

Not too bad, but Autofixture can help simplify the setup a good amount.   After importing the `Mutagen.Bethesda.Testing`, you should get all the goods letting you do this:
```cs
[Theory, MutagenAutoData]
public void MyTest(
    IFileSystem mockFs,
    MergeJsonExporter exporter)
{
    exporter.SomeCall(...);
    
    CheckSomeFileExistsAsExpected(mockFs);
}
```
Autofixture notices you want a IFileSystem for the test, and gives you a mock in-memory one by default.   It also notices you want a MergeJsonExporter for the test, which needs a IFileSystem.  It gives the same filesystem object it just gave you to the exporter.

Also, it can do random data pretty easily too
```cs
[Theory, MutagenAutoData]
public void MyOtherTest(
    ModKey someRandomModKey,
    int someRandomInt)
{
}
```

Helps just wire things up without all the boilerplate.  Lots of other nifty things it can do that I havent covered here

TLDR, can hopefully use the above systems to write some unit tests without too much headache.  Make a hand crafted  `MergeState` pass it to the exporter, then check the file system object for the output files and check expected content.

Comment (monyarm) [reply MutagenMerger.Lib/DI/Merger.cs:99]: On second thought, this kind of refactoring can wait till later. We can open an issue for it, and do it at a later time.
